### PR TITLE
[NTUSER] Avoid sending superfluous WM_MOUSEMOVE messages

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_I_Kill_Bugs patch to fix extra mouse moves generated when clicking._

JIRA issues: [CORE-8394](https://jira.reactos.org/browse/CORE-8394)
and the following:
[CORE-19423](https://jira.reactos.org/browse/CORE-19423)
[CORE-19422](https://jira.reactos.org/browse/CORE-19422)
[CORE-18529](https://jira.reactos.org/browse/CORE-18529)

## Proposed changes

_Add check so that mouse moves are not generated unless there is a changed in the mouse x or y position._

### CODE REVIEW

**Review by binarymaster — COMMENTED** [review]
I've renamed PR title to better express the meaning of it. Please take this into account when merging PR, because GitHub is applying initial PR title to the final commit message on merge (known GitHub bug).

**Comment by HBelusca** [comment]
Giving it a run in the testbots...

**Comment by Doug-Lyons** [comment]
> Giving it a run in the testbots...

Here is what I believe that it looks like on the testbots:

PR 6397 - Remove extra WM_mousemove on 0.4.15-dev-7444-gf51258a

VBox: https://reactos.org/testman/compare.php?ids=92737,92976
KVM:  https://reactos.org/testman/compare.php?ids=92738,92977

File: win32ss/user/ntuser/cursoricon.c
Comment (SergeGautherie) [inline win32ss/user/ntuser/cursoricon.c:267]: What about an early return?
```suggestion
    if (pt == gpsi->ptCursor)
        return TRUE;
```

Comment (HBelusca) [reply win32ss/user/ntuser/cursoricon.c:267]: not sure this compiles (equality test between _structures_)

**Comment by HBelusca** [comment]
> Here is what I believe that it looks like on the testbots:

Yes! I notice that `user32:TrackMouseEvent` seems to have some problems...
It would be of interest for some people to test some software under ROS with this PR.

File: win32ss/user/ntuser/cursoricon.c
Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: Right, modified.

**Comment by julenuri** [comment]
Also I have tested the [CORE-11775](https://jira.reactos.org/browse/CORE-11775) and every ticket related with sticky/non working are NOT solved yet.

**Comment by Doug-Lyons** [comment]
> Yes! I notice that `user32:TrackMouseEvent` seems to have some problems... It would be of interest for some people to test some software under ROS with this PR.

I saw this, but if you look at the VBox test without any modifications, it had 86 fails compared to the patched results having 90 fails. If you look at the most recent KVM tests you can see that 3 out of 5 had the same 86 fails:

https://reactos.org/testman/compare.php?ids=92991,92995,92998,93002,93005

File: win32ss/user/ntuser/cursoricon.c
Comment (Doug-Lyons) [reply win32ss/user/ntuser/cursoricon.c:267]: This would not take care of setting the last global pointer to server info (gpsi) statement here:
```
    gpsi->ptCursor = pt;
```
Though this statement could be moved above the 'if'. I prefer that there be fewer 'return's.

**Comment by Doug-Lyons** [comment]
> > Here is what I believe that it looks like on the testbots:
> 
> Yes! I notice that `user32:TrackMouseEvent` seems to have some problems... It would be of interest for some people to test some software under ROS with this PR.

There is a new patch by @I_Kill_Bugs ini https://jira.reactos.org/browse/CORE-11775 that seems to fix all of the failures in user32:TrackMouseEvent. I intend to create a PR for it within the next day or two. Thanks.

File: win32ss/user/ntuser/cursoricon.c
Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: > This would not take care of setting the last global pointer to server info (gpsi) statement

Why is that needed when the values are identical?

> I prefer that there be fewer 'return's.

ROS coding style prefers `return`s.

**Comment by HBelusca** [comment]
@Doug-Lyons Is that new patch you mention in the comment above be PR'ed since?

**Review by HBelusca — APPROVED** [review]

**Comment by Doug-Lyons** [comment]
> @Doug-Lyons Is that new patch you mention in the comment above be PR'ed since?

Yes. PR created here: https://github.com/reactos/reactos/pull/6426. Thanks.

**Review by JoachimHenze — APPROVED** [review]
Approved, because I do think this solution is rather straight-forward. Nevertheless I would encourage bisecting https://jira.reactos.org/browse/CORE-19423 and https://jira.reactos.org/browse/CORE-19422 because I do believe that this didn't always behave like that. I do assume it changed its behavior to the bad during 0.4.15-dev'ing.

**Comment by Doug-Lyons** [comment]
[reactosfanboy](https://github.com/secure/ViewProfile.jspa?name=reactosfanboy), I checked back to 0.4.0-20160208-r70702 and the problem did not exist at that time, so it looks like you were correct that is it a regression. I am not sure how much effort we want to spend in tracking this much further. I did check that the 0.4.14-release-99-g2f11139  and 0.4.14-dev-0-g43a6843 both DO show this problem, so it was not caused during our 0.4.15-dev'ing. Thanks.

**Comment by HBelusca** [comment]
> [reactosfanboy](https://github.com/secure/ViewProfile.jspa?name=reactosfanboy), I checked back to 0.4.0-20160208-r70702 and the problem did not exist at that time, so it looks like you were correct that it is a regression. I am not sure how much effort we want to spend in tracking this much further. I did check that the 0.4.14-release-99-g2f11139 and 0.4.14-dev-0-g43a6843 both DO show this problem, so it was not caused during our 0.4.15-dev'ing. Thanks.

Tracking regressions is important especially when it comes to core/vital modules.

**Comment by Doug-Lyons** [comment]
> > Tracking regressions is important especially when it comes to core/vital modules.

@HBelusca, based on your comment above I have spent many hours trying to find the guilty revision for this issue. But my results were without any significant degree of success. The problem shows up intermittently throughout its history from my testing. Some VM's do not produce the issue at all. Please see my more detailed comments in https://jira.reactos.org/browse/CORE-19422. Sorry I was not able to help more. I still feel that this PR has merit and should be committed because it does fix some VM's.

File: win32ss/user/ntuser/cursoricon.c
Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: Please, could you answer my question?

Comment (Doug-Lyons) [reply win32ss/user/ntuser/cursoricon.c:267]: I have reviewed the "Programming Guidelines" here https://reactos.org/wiki/Programming_Guidelines.
I saw no reference to "ROS coding style prefers ```return ```s.
My preference is for a single return point at the end of the function. It is just my preference.

Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: The question is "Why is that needed when the values are identical?".

> no reference to "ROS coding style prefers `return `s."

Read https://reactos.org/wiki/Coding_Style#Control_structures

Comment (Doug-Lyons) [reply win32ss/user/ntuser/cursoricon.c:267]: @SergeGautherie, Thank you for pointing me to the reference that you mentioned previously. I see that the reason for this recommendation is to reduce the indentation levels and I agree that if this code were modified as you suggest, we could remove the open brace (Line 268) and close brace (Line 275) with having only added one line with the ```return TRUE;```. I see your point here. Thanks again.
Let's leave it up to @HBelusca or another developer based on whoever finally decides to commit this as to what direction they would like to have this code take.

Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: You have not answered the question about old line 275 yet.

**Comment by jeditobe** [comment]
would be nice to see this finally merged ^)

**Review by katahiromz — APPROVED** [review]

**Comment by Doug-Lyons** [comment]
> I've renamed PR title to better express the meaning of it. Please take this into account when merging PR, because GitHub is applying initial PR title to the final commit message on merge (known GitHub bug).

@binarymaster, It would be nice if you could do this commit (it has 3 approvals now) and fix the commit message as you have noted here. Thanks.

**Comment by Doug-Lyons** [comment]
@binarymaster, Thanks for the commit and fixing the commit message. @JoachimHenze, @HBelusca, and @katahiromz, Thanks for your approvals and support.

File: win32ss/user/ntuser/cursoricon.c
Comment (SergeGautherie) [reply win32ss/user/ntuser/cursoricon.c:267]: Submitted as  #6770.

**Pull request merged**
